### ISSUE TITLE
Frequent NaN Loss During Training

### ISSUE THREAD

**Issue by Alvin-ne** [Graph-and-Geometric-Learning/HyperCore#2]
Hi HyperCore team,

I've been encountering frequent NaN loss issues while training models using the HyperCore library, which has hindered my training process. I've tried several mitigation strategies but haven't been able to resolve the problem. Here's a detailed breakdown:

During model training, the loss frequently becomes NaN. This occurs consistently across different hyperbolic models I've experimented with.

To address the NaN loss, I've tried the following approaches, but none have resolved the issue:
Using Riemannian Adam optimizer for optimizing manifold.c parameters.
Reducing learning rate to a very low value (lr = 5e-6) and weight decay (weight_decay = 1e-6).
Implementing learning rate schedulers (Linear warmup & Cosine decay).
Gradient clipping with max_grad = 1.0.

Could you provide guidance on potential causes of frequent NaN loss in HyperCore training and recommend additional strategies to mitigate this issue? For example, are there specific manifold parameter settings, stabilization techniques, or layer configurations that are more robust to numerical instability?

Let me know if you need more details about my model architecture, training setup, or specific modules used.

Thanks in advance for your help!

**Comment by Hellolder** [comment]
@Alvin-ne

NaN loss in hyperbolic manifold training is a known numerical stability issue. Based on your description, here are the likely causes:

## Root Cause Analysis

### 1. Unconstrained Curvature Parameter Optimization

The curvature parameter `c` is critical. If `c` grows unbounded during optimization:
- Exponential map `exp_c(x)` becomes undefined at ||x|| >= 1/sqrt(c)
- Logarithmic map `log_c(x)` produces complex numbers or NaN
- Hyperbolic distance calculations overflow

**Solution: Constrain `c` during optimization**

```python
# Method 1: Hard clamping
with torch.no_grad():
    manifold.c.data.clamp_(1e-6, 1.0)

# Method 2: Reparameterize c = softplus(c_raw)
# Optimize only c_raw, ensuring positivity
```

### 2. Embedding Norm Exceeds Manifold Domain

For Poincare Ball, embeddings must satisfy ||x|| < 1/sqrt(c). Near the boundary:
- Metric tensor g_x = 4/(1-c||x||^2)^2 diverges
- Gradient computation produces NaN

**Solution: Project embeddings back to valid domain**

```python
def project_to_poincare_ball(x, c, epsilon=1e-3):
    max_norm = (1.0 - epsilon) / torch.sqrt(c)
    norm = torch.norm(x, dim=-1, keepdim=True)
    return x * torch.min(max_norm / norm, torch.ones_like(norm))

# Call after each optimization step
x.data = project_to_poincare_ball(x.data, c)
```

### 3. Inadequate Gradient Clipping

Hyperbolic gradient norms differ from Euclidean. Standard clipping may be insufficient.

**Solution: Use more aggressive Riemannian gradient clipping**

```python
torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=0.5)
```

## Additional Recommendations

### Optimizer Choice
Consider **Riemannian AdamW** instead of Adam. Weight decay handling differs in hyperbolic space. See: *Riemannian Adaptive Optimization Methods* (ICLR 2019).

### Learning Rate
`lr=5e-6` may be too small, causing convergence to pathological regions. Try starting at `lr=1e-4` with warmup (500-1000 steps linear increase).

### Numerical Precision
If using **AMP (Automatic Mixed Precision)**, disable it. Hyperbolic functions are sensitive to FP16 precision.

## Debugging Script

```python
def debug_nan(model, c):
    for name, param in model.named_parameters():
        if param.grad is not None:
            if torch.isnan(param.grad).any():
                print(f"[NaN] {name}")
            grad_norm = param.grad.norm().item()
            if grad_norm > 10:
                print(f"[Large Grad] {name}: {grad_norm:.4f}")
    
    for name, module in model.named_modules():
        if hasattr(module, 'weight') and module.weight is not None:
            norm = module.weight.norm().item()
            max_allowed = 0.99 / torch.sqrt(c).item()
            if norm > max_allowed * 0.9:
                print(f"[Warning] {name}: norm={norm:.4f}")
```

## Follow-up Questions

To diagnose further:
1. Which manifold model? (Poincare Ball / Hyperboloid / Klein / Lorentz)
2. Model size?
3. Dataset type and scale?
4. Full training config (batch size, sequence length)?
5. At which step does NaN first appear?
